pkg/sql/rollback.go

```diff
@@ -123,58 +125,102 @@ func DeleteTableWithPredicate(
 	ctx context.Context,
 	db *kv.DB,
 	codec keys.SQLCodec,
+	distSender *kvcoord.DistSender,
 	table catalog.TableDescriptor,
 	predicates roachpb.DeleteRangePredicates,
 	batchSize int64,
 ) error {
 
 	log.Infof(ctx, "deleting data for table %d with predicate %s", table.GetID(), predicates.String())
-	tableKey := codec.TablePrefix(uint32(table.GetID()))
-	spans := []roachpb.Span{{Key: tableKey, EndKey: tableKey.PrefixEnd()}}
-
-	// TODO(msbutler): Currently, the distsender will split the predicate based
-	// delete range request into a requests for each range, and process them
-	// sequentially. Pre-split requests up using a rangedesc cache and run batches
-	// in parallel (since we're passing a key limit, distsender won't do its usual
-	// splitting/parallel sending to separate ranges).
-	for len(spans) != 0 {
-
-		span := spans[0]
-		admissionHeader := roachpb.AdmissionHeader{
-			Priority:                 int32(admissionpb.BulkNormalPri),
-			CreateTime:               timeutil.Now().UnixNano(),
-			Source:                   roachpb.AdmissionHeader_FROM_SQL,
-			NoMemoryReservedAtSource: true,
+	tableKey := roachpb.RKey(codec.TablePrefix(uint32(table.GetID())))
+	tableSpan := roachpb.RSpan{Key: tableKey, EndKey: tableKey.PrefixEnd()}
+
+	// TODO (msbutler): tune these
+	rangesPerBatch := int64(100)
+	numworkers := 4
+
```

Comment:
Currently, I buffer up all the partitioned spans before issuing requests, which could be memory intensive for a big enough table. I tried putting this initial span sending into a separate go routine to reduce the span buffer size and to allow span partition creation and DeleteRange requests to occur simultaneously, but I couldn't think of a good design after an hour.  If folks think this is a big problem, would be great to pair program on this. I don't have much experience working channels.

Reply:
We don't really need a separate goroutine, we can use the current one. First create the channel, then spin off the workers (who will wait on the channel), and then just feed the spans to the channels in the for loop (where you currently append to `initialSpans`). When done, close the channel and call `Wait()` on the workers.

Reply:
Aha! Thanks for the insight! Code looks much cleaner now.

Aside: I based my initial approach on some examples in the bulk processors, one of which uses the antipattern of a go thread sending and receiving on the same channel. I wonder if that should change 🤔

Reply:
>Aside: I based my initial approach on some examples in the bulk processors, one of which uses the antipattern of a go thread sending and receiving on the same channel. I wonder if that should change 🤔

Right. It's sometimes warranted, but usually it's a good idea to avoid it.